evproc/processor.py

```diff
@@ -308,11 +311,16 @@ def process(self, ev):
                   ``StopProcessing`` exception raised by a processor.
         """
 
-        # Walk through all the processors and process the event
-        for proc in self._get_procs(ev):
-            try:
-                proc(ev)
-            except exc.StopProcessing as ex:
-                return ex.retval
+        event_deque = collections.deque([ev])
+        while event_deque:
+            ev = event_deque.popleft()
+            # Walk through all the processors and process the event
+            for proc in self._get_procs(ev):
+                try:
+                    additional_events = proc(ev)
+                except exc.StopProcessing as ex:
+                    return ex.retval
```

Comment:
The simplest way to implement this, I think, is that if a StopProcessing exception contains a retval (which may be None), then Processor.process() will immediately return that retval and ignore any remaining events in the queue; if a StopProcessing exception does not contain a retval, then Processor.process() will stop processing the current event (ignoring any remaining handlers for that event) and proceed to process the next event in the queue. This would be independent of whether or not the event is the first one.

I think this would be backwards compatible, and would be rather intuitive, I think.